fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -616,18 +650,36 @@ def _shard_parameters_(self) -> None:
             if not p._is_sharded:
                 p._is_sharded = False
                 self.numel_padded_per_param.append(0)
+                # Insert tensor into the SSD buffer and free parameter storage.
+                if self.ssd_offload:
+                    p._shard_size = p.data.size()  # type: ignore
+                    p._handle = self.ssd_buffer.insert(p.data)  # type: ignore
+                    free_storage_(p.data)
                 continue
             p._is_sharded = True
 
             # Replace p.data with the relevant shard.
-            orig_data = p.data
-            p.data, num_padded = self._get_shard(p.data)
-            self.numel_padded_per_param.append(num_padded)
-            free_storage_(orig_data)
+            if self.ssd_offload:
+                orig_data = p.data
+                p.data, num_padded = self._get_shard(p.data)
+                p._shard_size = p.data.size()  # type: ignore
+                # Insert tensor into the SSD buffer and free parameter storage.
+                p._handle = self.ssd_buffer.insert(p.data)  # type: ignore
+                del orig_data
+                self.numel_padded_per_param.append(num_padded)
+                free_storage_(p.data)
+            else:
+                orig_data = p.data
+                p.data, num_padded = self._get_shard(p.data)
+                self.numel_padded_per_param.append(num_padded)
+                free_storage_(orig_data)
 
-            p._is_sharded = True
         assert len(self.numel_padded_per_param) == len(self.params)
 
+        # Move SSD buffer to disk.
+        if self.ssd_offload:
+            self.ssd_buffer.to_disk()
```

Comment:
Can you please restructure this part to:
```
< original code>

if ssd_offload:
   <ssd offload code>
```
This way, the impact of offload code is easier to see and to separate for debugging etc.

Reply:
if reserving the order is easier, that's fine too, like:

```
if ssd:
   <ssd code>
   return

<original code>
```

A bit of duplicated code is fine too but I see the code above is mostly separately anyway. so there shouldn't be any duplicates.

Reply:
restructured it within the is_sharded vs not blocks. Let me know if this is what you had in mind as well.